File: timeout/golang_fiberv3/repositories/test1_repository.go
package repositories

import (
	"context"
	modelentities "note-golang-fiberv3-timeout/models/entities"

	"github.com/jmoiron/sqlx"
)

type Test1Repository interface {
	Sleep(tx *sqlx.Tx, ctx context.Context, second int) (result string, err error)
	SleepWithDb(db *sqlx.DB, ctx context.Context, second int) (result string, err error)
	Create(tx *sqlx.Tx, ctx context.Context, test1 modelentities.Test1) (rowsAffected int64, err error)
	CreateWithDb(db *sqlx.DB, ctx context.Context, test1 modelentities.Test1) (rowsAffected int64, err error)
}

type test1Repository struct {
}

func NewTest1Repository() Test1Repository {
	return &test1Repository{}
}

func (repository *test1Repository) Sleep(tx *sqlx.Tx, ctx context.Context, second int) (result string, err error) {
	err = tx.GetContext(ctx, &result, `SELECT PG_SLEEP($1);`, second)
	return
}

func (repository *test1Repository) SleepWithDb(db *sqlx.DB, ctx context.Context, second int) (result string, err error) {
	err = db.GetContext(ctx, &result, `SELECT PG_SLEEP($1);`, second)
	return
}

func (repository *test1Repository) Create(tx *sqlx.Tx, ctx context.Context, test1 modelentities.Test1) (rowsAffected int64, err error) {
	result, err := tx.ExecContext(ctx, `INSERT INTO test1(test) VALUES($1);`, test1.Test)
	if err != nil {
		return
	}
	return result.RowsAffected()
}

func (repository *test1Repository) CreateWithDb(db *sqlx.DB, ctx context.Context, test1 modelentities.Test1) (rowsAffected int64, err error) {
	result, err := db.ExecContext(ctx, `INSERT INTO test1(test) VALUES($1);`, test1.Test)
	if err != nil {
		return
	}
	return result.RowsAffected()
}

File: timeout/rust/src/repositories/test1_repository.rs
use sqlx::{Postgres, Transaction};

use crate::models::entities::test1::Test1;

pub trait Test1Repository {
    async fn create_without_tx(&self, pool: &sqlx::PgPool, test1: &Test1) -> Result<(), sqlx::Error>;
    async fn sleep_without_tx(&self, pool: &sqlx::PgPool) -> Result<(), sqlx::Error>;
    async fn create_with_tx(&self, tx: &mut Transaction<'_, Postgres>, test1: &Test1) -> Result<(), sqlx::Error>;
    async fn sleep_with_tx(&self, tx: &mut Transaction<'_, Postgres>) -> Result<(), sqlx::Error>;
}

pub struct Test1RepositoryImpl{}

impl Test1RepositoryImpl {
    pub fn new() -> Test1RepositoryImpl {
        Test1RepositoryImpl {  }
    }
}

impl Test1Repository for Test1RepositoryImpl {
    async fn create_without_tx(&self, pool: &sqlx::PgPool, test1: &Test1) -> Result<(), sqlx::Error> {
        let _ = sqlx::query("INSERT INTO test1(test) VALUES($1);")
            .bind(&test1.test)
            .execute(pool)
            .await?;
        Ok(())
    }

    async fn sleep_without_tx(&self, pool: &sqlx::PgPool) -> Result<(), sqlx::Error> {
        let _ = sqlx::query("SELECT pg_sleep(3);")
            .execute(pool)
            .await?;
        Ok(())
    }

    async fn create_with_tx(&self, tx: &mut Transaction<'_, Postgres>, test1: &Test1) -> Result<(), sqlx::Error> {
        let _ = sqlx::query("INSERT INTO test1(test) VALUES($1);")
        .bind(&test1.test)
        .execute(&mut **tx)
        .await?;
        Ok(())
    }

    async fn sleep_with_tx(&self, tx: &mut Transaction<'_, Postgres>) -> Result<(), sqlx::Error> {
        let _ = sqlx::query("SELECT pg_sleep(3);")
        .execute(&mut **tx)
        .await?;
        Ok(())
    }
}